rllib/execution/parallel_requests.py

```diff
@@ -182,3 +183,224 @@ def wait_asynchronous_requests(
+        *,
+        fn_args: List[Any] = None,
+        fn_kwargs: Dict[str, Any] = None,
+        for_all_workers: bool = False,
+    ) -> None:
+        """Submit a remote function call schedule on available workers.
+
+        Args:
+            remote_fn: The remote function to call
+            fn_args: The arguments to pass to the remote function
+            fn_kwargs: The keyword arguments to pass to the remote function
+            for_all_workers: If True, submit this request to the queue for as many
+                number of workers that there are. If False, submit the request without
+                replicating it.
+
+        """
+        if fn_args is None:
+            fn_args = []
+        if fn_kwargs is None:
+            fn_kwargs = {}
+        call = (remote_fn, fn_args, fn_kwargs)
+        if for_all_workers:
+            for _ in range(self._num_workers):
+                self._call_queue.put(call)
+        else:
+            self._call_queue.put(call)
+        self._run()
+
+    def _run(self) -> int:
+        """Run submitted requests on available workers"""
+        num_requests_launched = 0
+        available_actor = self._get_available_actor()
+        while available_actor and self._call_queue.qsize() > 0:
+            remote_fn, fn_args, fn_kwargs = self._call_queue.get()
+            req = available_actor.apply.remote(remote_fn, *fn_args, **fn_kwargs)
+            self._remote_requests_in_flight[available_actor].add(req)
+            self._pending_remotes.append(req)
+            self._pending_to_actor[req] = available_actor
+            available_actor = self._get_available_actor()
+            num_requests_launched += 1
+        return num_requests_launched
+
+    def get_ready_requests(self) -> Dict[ActorHandle, List[Any]]:
+        """Get requests that are ready to be returned
+
+        Returns:
+            A dictionary of actor handles to lists of returns from tasks that were
+             previously submitted to this actor pool that are now ready to be returned.
+             If return_object_refs
+
+        """
+        ready_requests_dict = defaultdict(list)
+        ready_requests, self._pending_remotes = ray.wait(
+            self._pending_remotes,
+            timeout=self._ray_wait_timeout_s,
+            num_returns=len(self._pending_remotes),
+        )
+        if not self._return_object_refs:
+            objs = ray.get(ready_requests)
+        else:
+            objs = ready_requests
+        for req, obj in zip(ready_requests, objs):
+            actor = self._pending_to_actor[req]
+            self._remote_requests_in_flight[actor].remove(req)
+            ready_requests_dict[actor].append(obj)
+            if actor in self._unavailable_workers:
+                self._available_workers.append(actor)
+                self._available_workers_set.add(actor)
+                self._unavailable_workers.remove(actor)
+            del self._pending_to_actor[req]
+        del ready_requests
```

Comment:
in general I feel like we should simplify the states we track.
tracking a lot of states, and making sure everything is in-sync and correct is bug prone.